devtools/generate_supported.py

```diff
@@ -0,0 +1,253 @@
+            print("##NEW##")
+            print(supported_text)
+
+        new_contents = contents[:start_index]
+        end_contents = contents[end_index:]
+        new_contents.append(supported_text)
+        new_contents.extend(end_contents)
+
+        with open(filename, "w") as f:
+            new_contents_text = "".join(new_contents)
+            f.write(new_contents_text)
+        return True
+    return False
+
+
+def _supported_summary(supported):
+    return _supported_text(
+        supported,
+        "### Supported $brand$auth devices\n\n$types\n",
+        "- **$type_**: $models\n",
+    )
+
+
+def _supported_detail(supported):
+    return _supported_text(
+        supported,
+        "## $brand devices\n\n$preamble\n\n$types\n",
+        "### $type_\n\n$models\n",
+        "- **$model**\n$versions",
+        "  - Hardware: $hw$region / Firmware: $fw$auth_flag\n",
+    )
+
+
+def _supported_text(
+    supported, brand_template, types_template, model_template="", version_template=""
+):
+    brandt = Template(brand_template)
+    typest = Template(types_template)
+    modelt = Template(model_template)
+    versst = Template(version_template)
+    brands = ""
+    version: SupportedVersion
+    for brand, types in supported.items():
+        preamble_text = (
+            "<sup>*</sup>&nbsp; Kasa model requires authentication"
+            if brand == "kasa"
+            else "All Tapo devices require authentication"
+        )
+        brand_text = brand.capitalize()
+        brand_auth = r"<sup>\*</sup>" if brand == "tapo" else ""
+        types_text = ""
+        for supported_type, models in sorted(
+            # Sort by device type order in the enum
+            types.items(),
+            key=lambda st: list(SupportedDeviceType).index(st[0]),
+        ):
+            models_list = []
+            models_text = ""
+            for model, versions in sorted(models.items()):
+                auth_count = 0
+                versions_text = ""
+                for version in sorted(versions):
+                    region_text = f" ({version.region})" if version.region else ""
+                    auth_count += 1 if version.auth else 0
+                    vauth_flag = r"<sup>\*</sup>" if version.auth else ""
+                    vauth_flag = "" if brand == "tapo" else vauth_flag
+                    if version_template:
+                        versions_text += versst.substitute(
+                            hw=version.hw,
+                            fw=version.fw,
+                            region=region_text,
+                            auth_flag=vauth_flag,
+                        )
+                auth_flag = (
+                    r"<sup>\*</sup>"
+                    if auth_count == len(versions)
+                    else r"<sup>\*\*</sup>"
+                    if auth_count > 0
+                    else ""
+                )
+                auth_flag = "" if brand == "tapo" else auth_flag
```

Comment:
This is pretty much the same as `vauth_flag`, maybe separate the handling into a helper function to avoid repetition?